PhysicsTools/NanoAOD/plugins/NPUTablesProducer.cc

```diff
@@ -24,24 +27,51 @@ class NPUTablesProducer : public edm::global::EDProducer<> {
         void produce(edm::StreamID id, edm::Event& iEvent, const edm::EventSetup& iSetup) const override {
             auto npuTab  = std::make_unique<nanoaod::FlatTable>(1, "Pileup", true);
 
-            edm::Handle<std::vector<PileupSummaryInfo> > npuInfo;
+            
+	    edm::Handle<edm::View<pat::PackedGenParticle> >        genParticles;
+	    double refpvz = -1000.0;
+	    if ( iEvent.getByToken(genTag_, genParticles) ) {
+	      for ( auto genIt = genParticles->begin(); genIt != genParticles->end(); ++genIt ) {
+		if ( genIt->statusFlags().isHardProcess() ) {
```

Comment:
I tested the code and it seems that none of the `packedGenParticles` satisfy this condition. What I see is that the variable `refpvz` will always have a value of `-1000`, which in turn means that the `zbin` variable in `fillNPUObjectTable()` refers to the last element of `vz_`. Provided that the element beyond `vz_` array is 0, as is usually the case in practice (even though it could be any number), the denominator on line 74 is `-12 * 20 = -240`.
I think that `packedGenParticles` (nor `prunedGenParticles`) are not suitable for this logic by construction:
https://github.com/cms-nanoAOD/cmssw/blob/078033feb18754c58b1222219493cb910975ad71/PhysicsTools/PatAlgos/plugins/PATPackedGenParticleProducer.cc#L99-L121
What I suggest is to instead smuggle `offlineSlimmedPrimaryVertices` into this class, and read `refpvz` with:
https://github.com/cms-nanoAOD/cmssw/blob/078033feb18754c58b1222219493cb910975ad71/PhysicsTools/NanoAOD/plugins/VertexTableProducer.cc#L135